File: 0312_Burst Balloons/BurstBalloonsRecursion.java
class Solution {
    private int[][] memo;
    private int[] vals;
    public int maxCoins(int[] nums) {
      int n = nums.length;
      vals = new int[n + 2];
      vals[0] = vals[n + 1] = 1;
      for (int i = 0; i < n; ++i){
        vals[i + 1] = nums[i];
      } 
      memo = new int[n + 2][n + 2];
      return maxCoins(1, n);
    }
    
    private int maxCoins(int i, int j) {
      if (i > j) return 0;
      if (i == j) return vals[i - 1] * vals[i] * vals[i + 1];
      if (memo[i][j] > 0) return memo[i][j];
      int most = 0;
      for (int k = i; k <= j; ++k)
        most = Math.max(most, maxCoins(i, k - 1) + vals[i - 1] * vals[k] * vals[j + 1] + maxCoins(k + 1, j));
      memo[i][j] = most;
      return memo[i][j];
    }
  }

File: 0312_Burst Balloons/BurstBalloonsRecursion.cs
public class Solution {
    private int[] vals;
    private int[,] mem;
    public int MaxCoins(int[] nums) {
        
        mem = new int[nums.Length+2, nums.Length+2];
        vals = new int[nums.Length+2];
        vals[0] = 1;
        vals[vals.Length -1] = 1;
        
        for(int i=0;i<nums.Length;i++)
        {
            vals[i+1] = nums[i];
        }
        
        return MaxCoins(1,nums.Length);
        
    }
    
    private int MaxCoins(int i, int j)
    {
        if(i > j) return 0;
        if(i == j) return vals[i-1] * vals[i] * vals[i+1];
        if(mem[i,j] > 0) return mem[i,j];
        
        var most = 0;
        for(int k = i;k<=j;k++)
        {
            most = Math.Max(most, MaxCoins(i,k-1) + vals[i-1] * vals[k] * vals[j+1] + MaxCoins(k+1,j));
        }
        
        mem[i,j] = most;
        
        return most;
    }
}